request-server.go

```diff
@@ -57,11 +63,17 @@ func (rs *RequestServer) nextRequest(r *Request) string {
 	defer rs.openRequestLock.Unlock()
 	rs.handleCount++
 	handle := strconv.Itoa(rs.handleCount)
-	rs.openRequests[handle] = r
+
+	ctx, cancel := context.WithCancel(context.Background())
+	r = r.WithContext(ctx)
```

Comment:
Shouldn’t this be the responsibility of the person making the request?

What if we don’t want every request to have a context?

Default behavior should be that contexts are not added in, if they’re not explicitly given by the user.

Reply:
This is where I somewhat disagree, I'd say it makes sense to attach a cancellable context at the layer that is creating these requests and knows when the requests are done and thus can cancel the attached context. For example in the net/http Server implementation:

https://github.com/golang/go/blob/release-branch.go1.9/src/net/http/server.go#L1730

They setup a cancelFunc and context at the time of accepting the TCP connection and cancel it once that connection has completed.

Reply:
Hm… but you notice the design there is to build a context, with a cancelFunc, then defer the cancel… here, it seems that we stuff the context, and cancel into a structure, and who knows where it gets taken care of.

Reply:
That's a good point, but I'm not sure if we can set up a similar flow given the way packetWorker switches off specific sftp packet types. 

Here I'm sort of relying on nextRequest getting called first and eventually a closeRequest, but I'm unsure if the sshFxpClosePacket is guaranteed to be received even if there's a non-clean ending to the underlying connection.